File: lib/js/markov.js
/** Markov chain — JavaScript implementation. Pure functions, no state. */
const Mark = {
  build(notes) {
    const table = {};
    for (let i = 0; i < notes.length - 1; i++) {
      const from = notes[i];
      const to = notes[i + 1];
      table[from] = table[from] || {};
      table[from][to] = (table[from][to] || 0) + 1;
    }
    // Normalize
    for (const from of Object.keys(table)) {
      const total = Object.values(table[from]).reduce((a, b) => a + b, 0);
      for (const to of Object.keys(table[from])) {
        table[from][to] /= total;
      }
    }
    return table;
  },

  generate(table, start, length) {
    const seq = [start];
    let current = start;
    for (let i = 0; i < length - 1; i++) {
      const targets = table[current];
      if (targets) {
        const r = Math.random();
        let cumulative = 0;
        for (const [next, prob] of Object.entries(targets)) {
          cumulative += prob;
          if (r <= cumulative) {
            seq.push(Number(next));
            current = Number(next);
            break;
          }
        }
      } else {
        seq.push(current);
      }
    }
    return seq;
  }
};

if (typeof module !== 'undefined') module.exports = Mark;
if (require.main === module) {
  const training = [60, 62, 64, 65, 67, 65, 64, 62];
  const table = Mark.build(training);
  console.log('Training:', training);
  console.log('Generated:', Mark.generate(table, 60, 16));
  console.log('Table:', JSON.stringify(table, null, 2));
}

File: lib/go/markov.go
package main

import (
	"fmt"
	"math/rand"
)

type TransitionTable map[uint8]map[uint8]float64

func BuildTransitions(notes []uint8) TransitionTable {
	table := make(TransitionTable)
	for i := 0; i < len(notes)-1; i++ {
		from := notes[i]
		to := notes[i+1]
		if table[from] == nil {
			table[from] = make(map[uint8]float64)
		}
		table[from][to]++
	}
	for _, targets := range table {
		var total float64
		for _, count := range targets {
			total += count
		}
		for note, count := range targets {
			targets[note] = count / total
		}
	}
	return table
}

func Generate(table TransitionTable, start uint8, length int) []uint8 {
	seq := make([]uint8, 0, length)
	seq = append(seq, start)
	current := start
	for i := 0; i < length-1; i++ {
		if targets, ok := table[current]; ok {
			r := rand.Float64()
			var cumulative float64
			for next, prob := range targets {
				cumulative += prob
				if r <= cumulative {
					seq = append(seq, next)
					current = next
					break
				}
			}
		} else {
			seq = append(seq, current)
		}
	}
	return seq
}

func main() {
	training := []uint8{60, 62, 64, 65, 67, 65, 64, 62}
	table := BuildTransitions(training)
	result := Generate(table, 60, 16)
	fmt.Println("Training:", training)
	fmt.Println("Markov:", result)
}
